osbenchmark/worker_coordinator/worker_coordinator.py

```diff
@@ -198,6 +198,171 @@ def __init__(self, metrics, next_task_scheduled_in):
+    NEUTRAL = "neutral"
+    SCALING_DOWN = "scaling_down"
+    SLEEP = "sleep"
+    SCALING_UP = "scaling_up"
+
+class FeedbackActor(actor.BenchmarkActor):
+    POST_SCALEDOWN_SECONDS = 30
+    WAKEUP_INTERVAL = 1
+
+    def __init__(self):
+        super().__init__()
+        self.logger = logging.getLogger(__name__)
+        self.state = FeedbackState.SCALING_UP
+        self.messageQueue = queue.Queue()
+        self.shared_client_states = {}
+        self.total_active_client_count = 0
+        self.total_client_count = 0
+        self.sleep_start_time = None
+        self.last_error_time = None
+        self.last_scaleup_time = None
+
+    def handle_state(self):
+        current_time = time.time()
+
+        if self.state == FeedbackState.SLEEP:
+            # Check if we've slept for long enough to return to a normal state
+            if current_time - self.sleep_start_time >= self.POST_SCALEDOWN_SECONDS:
+                self.logger.info("Feedback Actor's sleep period complete, returning to NEUTRAL state")
+                self.state = FeedbackState.NEUTRAL
+                self.sleep_start_time = None
+            return
+
+        if self.messageQueue.qsize() > 0:
+            self.logger.info("Feedback Actor has received an error message, scaling down...")
+            self.state = FeedbackState.SCALING_DOWN
+            self.scale_down()
+            self.logger.info("Clients scaled down. Number of active clients: %d", self.total_active_client_count)
+
+        if self.state == FeedbackState.NEUTRAL:
+            # Check if we've waited long enough since the last scaledown
+            if current_time - self.last_error_time >= self.POST_SCALEDOWN_SECONDS:
+                if (self.last_scaleup_time is None or
+                    current_time - self.last_scaleup_time >= self.WAKEUP_INTERVAL):
+                    self.logger.info("no errors in the last 30 seconds, scaling up")
+                    self.state = FeedbackState.SCALING_UP
+                    self.scale_up()
+                    self.last_scaleup_time = current_time
+            else:
+                self.logger.info("Cluster has errored too recently, waiting before scaling up")
+
+    def receiveMsg_ClusterErrorMessage(self, msg, sender):
+        self.last_error_time = time.time()
+        self.logger.info("Feedback actor has recevied an error message.")
+        if self.state == FeedbackState.SCALING_DOWN:
+            self.logger.info("Already scaling down, ignoring error")
+            return
+        elif self.state == FeedbackState.SLEEP:
+            self.logger.info("In sleep mode, ignoring error")
+            return
+        elif self.state == FeedbackState.NEUTRAL:
+            # Add error message to queue
+            try:
+                self.messageQueue.put(msg)
+            except queue.Full:
+                self.logger.error("Queue is full - message dropped: %s", msg)
+
+    def scale_down(self, scale_down_percentage=0.10):
```

Comment:
same here and with other execution methods, move them below all the message handling logic.